### ISSUE TITLE
Evolving Generative Adversarial Networks to improve image steganography method in python

### ISSUE THREAD

**Issue by hosseinxh** [hosseinxh/Evolving-Generative-Adversarial-Networks-to-improve-image-steganography#1]
python code for the following steps (explanation and structure of the model):

Generator
The generator consists of a convolutional neural network (CNN) which takes a fixed-size image as input and returns an image with the same dimensions.

This network receives a cover image as input (i.e, an image without a steganographic message). To add some random factor to the image modification, another channel composed of Gaussian noise with mean 0 and variance 1 is included. The CNN input therefore will be composed of 2 channels and will return 1 channel representing an imagen preserving the same dimensions as the input image.
Given that the goal is to return an image as similar as possible to the one used as input, an scheme that enlarges and reduces the size of the input image is used. First the input image dimensions are increased using transpose convolutions. Then, once the image has been enlarged, a mirror architecture is used to reduce it to its original size. This avoids data loss while it requires more computational resources.
As one can see in Fig. 1 the proposed architecture has 5 transpose convolutional layers which increase the input size from 256 × 256 to 528 × 528. In addition, the number of filters increase from 2 to 16 in the first layer and to 32 in the third layer. After this, there are six convolutional layers which decrease the image size to 256 × 256, its original size. Leaky Relu is used as activation function in all layers except in the last one, which use tanh activation function, as recommended in the literature (Radford et al., 2015). Finally, while layers 6 and 7 have kernel size 5, the others have kernel size 3.

Fig. 1. Diagram showing the layers composing the generator and discriminator modules of the Generative Adversarial Network designed(the first picture 1.png)

Discriminator
For the discriminator, a steganalysis model based on a CNN architecture from the state of the art was selected. The chosen model is yenet (Ye et al., 2017). This CNN model will be pre-trained using the same dataset that in all the experiments performed in this work. This will allow to extend this work to other steganalysis models.

Steganographic module
Once the generator has modified the input image, the steganographic message can be embedded. For that purpose, it is necessary to integrate a new module between the generator and the discriminator running in the same GPU computing environment. The GAN model was implemented using Tensorflow, so the module will implement the steganographic algorithm using Tensors.

Training strategy
It must be noted that the generator is trained pursuing 2 different objectives. The first one is to keep the modified image as similar to the original image as possible. To that end, cosine similarity measure is used as loss value.

The cosine similarity was chosen since, in contrast to other metrics, its provides a reliable metric based on orientations, and that can be safely applied where magnitudes are not an issue.

![1](https://github.com/hosseinxh/Evolving-Generative-Adversarial-Networks-to-improve-image-steganography/assets/71155677/f86d6f33-587b-4136-bb0a-6fd07a611dad)

if it's continues keep going and if it's completed for now, follow the rest:

In the scenario proposed in this research, the comparison is performed between images where features (i.e. a pixel) is defined by a bounded range of values. Besides, cosine similarity has already been successfully integrated in GAN architectures due to its flexibility and efficiency (Zhan et al., 2022), but also in other tasks such as contrastive learning (Chen et al., 2022). This function is shown in Eq.(2)
<math xmlns="http://www.w3.org/1998/Math/MathML">
  <mrow is="true">
    <mi is="true">S</mi>
    <mi is="true">I</mi>
    <mi is="true">M</mi>
    <mrow is="true">
      <mo is="true">(</mo>
      <mi is="true">X</mi>
      <mo is="true">,</mo>
      <mi is="true">Y</mi>
      <mo is="true">)</mo>
    </mrow>
    <mo linebreak="goodbreak" is="true">=</mo>
    <mfrac is="true">
      <mrow is="true">
        <mrow is="true">
          <mo is="true">&#x3008;</mo>
          <mi is="true">X</mi>
          <mo is="true">,</mo>
          <mi is="true">Y</mi>
          <mo is="true">&#x3009;</mo>
        </mrow>
      </mrow>
      <mrow is="true">
        <mo is="true">&#x2016;</mo>
        <mi is="true">X</mi>
        <mo is="true">&#x2016;</mo>
        <mo linebreak="badbreak" is="true">&#x2217;</mo>
        <mo is="true">&#x2016;</mo>
        <mi is="true">Y</mi>
        <mo is="true">&#x2016;</mo>
      </mrow>
    </mfrac>
  </mrow>
</math>

The original image is considered as the label while the modified image is used as the predicted class. The loss value consists of a matrix with the same dimensions that the output (and input) matrix. This loss propagates through the generator updating the weights.

The second stage consists on training the generator to deceive the discriminator. Starting from the modified image, the first step involves embedding the steganographic message using the module explained above. Our goal is to generate images that, once the steganographic message is embedded, they are classified by the discriminator as cover. Once this classification is performed by the discriminator, the loss is computed. Then both networks, generator and discriminator, are concatenated and the loss propagates through the combined network as a standard CNN using the back-propagation algorithm and updating the network weights. However, discriminator weights are frozen so only the generator weights are updated. (see Fig. 2, 2.png).

Fig. 3. (3.png) Diagram describing the GAN model. The model steps are indicated by the numbers in the image. Images goes through 2 different ways: the path 1-2-6-7, which updates the generator using the cosine similarity loss; and the path 1-2-3-4-5-8, which updates the generator based on the discriminator loss. Steps 1–2 are shared and only performed once. The steps are: (1) the generator input is a 2 channel image consisting of a cover image and Gaussian noise, (2) the generator modifies the input image, (3) the steganographic module that embeds the steganographic message and label them as cover, (4) the stego images labelled as cover are classified by the discriminator, (5) the discriminator loss is computed using Adam optimiser, (6) the similarity loss is computed using the cosine similarity function between the modified and original images, (7) the generator weights are updated using the similarity loss, (8) the generator weights are updated using the discriminator loss.

remmember for dataset I want to use only 10 grayscale image with .pgm format

![2](https://github.com/hosseinxh/Evolving-Generative-Adversarial-Networks-to-improve-image-steganography/assets/71155677/42ab8e54-9d39-41a3-b53e-3760332ca058)

keep going and here is the rest:
4. Evolving GAN architectures for steganography
The second phase of this work consists on designing an algorithm based on evolutionary computation to find the best GAN architecture. There are some articles in the literature which take a similar approach, such as Evodeep (Martín et al., 2017, Martín et al., 2018), which implements a genetic algorithm to find the best architecture and hyper-parameters for CNNs. It evolves a population of networks modifying the number and hyper-parameter of layers in each individual. In this research, the designed method is a genetic algorithm which evolves a population of generators. Coral Reef-based optimisation has also be used to optimise the parameters of CNNs (Martín, Lara-Cabrera et al., 2020, Martín, Vargas et al., 2020).
4.1. Codification
The codification defines the architecture that can be generated by the algorithm. Its main goal is to find the architectures that best fit the problem (modify images to deceive the discriminator keeping the similarity to the original one as high as possible). However, the design of the population codification must be accomplished trying to keep the computational resources required as low as possible due to the computational complexity of the algorithm.

The codification will define CNNs with a similar structure to the generator proposed in the previous section. As shown in Fig. 4, the individuals are composed by 3 different groups. The first one consists on a group of transpose convolutional layers which enlarge the size of the image, then a group of convolutional layers that keep the same dimensions at the input and the output (called intermediate convolutions from now on), and finally a group of convolutional layers that decrease its size to its original size (named final convolutions in this work).
The generation of new individuals is guided by a very simple state machine (see Fig. 5). It is composed of only 2 states, the first one corresponds to the group of transpose convolutional layers and the second one to the intermediate group of convolutions. The final convolutional layers will be computed as a mirror of the transpose convolutional layers group, considering both the number of layers and their hyper-parameters.
Each state has a set of hyper-parameters that also evolve during the algorithm execution. Each one of the transpose convolutional layers (and the final convolutional layers) has the following hyper-parameters: kernel size, which can be 3 or 5; stride, which can be 1 or 2; the activation function, which can be reLU or Leaky ReLU; number of filters, which indicates the factor used to increase the number of filters by; and batch normalisation, which indicates if there is a batch normalisation layer after it. On the other side, the intermediate convolutional layers have the same hyper-parameters except for the stride which is always 1. The hyper-parameters of the final convolutional layers are the same that their corresponding mirror transpose convolutional layer.
In addition, the hyper-parameters are limited to prevent the individuals to growth in excess. The stride can only be 2 in one of the transpose convolutional layers, setting it to 1 in all the layers posterior to the first one with stride 2. This will prevent the individuals to achieve too high dimensions. The number of filters can be 1, 2 or 3, and the number of filters of the layer is computed multiplying the previous layer number of filters by this factor. However, it cannot exceed a certain value defined as a parameter of the genetic algorithm. All the layers that would exceed this limit will be set to it. Finally, the last layer of the network always has a single filter and its activation function is the tanh function, as its recommended in the literature (Radford et al., 2015)
![3](https://github.com/hosseinxh/Evolving-Generative-Adversarial-Networks-to-improve-image-steganography/assets/71155677/9577ef81-8a07-4fa7-82d9-5bff4597441f)

and the rest:
Crossover
The crossover operator combines 2 individuals to generate 2 new descendants. It is composed of 2 different stages. The first one is the external crossover, which performs a 1-point crossover between the layers of both parents. First, the number of layers of the new descendant is selected from the range of possible values (set as genetic algorithm parameter). Then the first point is randomly selected from the first parent from those who allow to generate valid individuals, i.e, those who allow to generate an individual from the selected number of layers. Then the cross point in the second parent is generated from those who generate validate individuals, i.e, from those which keep the selected number of layers and from the same group than in the first parent (transpose convolution or intermediate convolution).

The second stage is the internal crossover. It performs an uniform crossover between the hyper-parameters of all layers. The individual codification for which the final convolutions are computed from the group of transpose convolutions allow to generate always valid individuals.

Mutation
The mutation operator is performed after crossover and it is also composed of 2 stages. In the first one there is a probability called new layer probability, which is set as parameter of the genetic algorithm, of adding a new layer in a random location of the network. The location where the layer is added is selected from those which do not increase the individual size above the layer limit. It must be considered that if a transpose convolutional layer is added, then a convolutional layer is added in the group of final convolutions (the mirror layer).

The second stage consists on a random mutation of each one of the hyper-parameters of each one of the layers of the network. The mutation probability is set as a hyper-parameter of the genetic algorithm, and it is called from now on mutation probability.

Evaluation
The last component of the genetic algorithm is the individual evaluation. To evaluate an individual, a GAN where the generators corresponds to its codification must be implemented. Then, the GAN model is trained during a predefined number of epochs, set as a parameter of the genetic algorithm. The GAN is trained using the training set, and evaluated with the validation set. The discriminator loss and similarity loss obtained with the validation set are used as fitness value of the individual, which allow to select the best individuals between the old and new population.

Selection
The selection algorithm must choose between the individuals of the current population and their descendants. According to its criterion, a number of individuals equal to the population size must be selected. The selection algorithm used in this work is called NSGA-II or Non dominated sorting genetic algorithm (Deb, Pratap, Agarwal, & Meyarivan, 2002). This algorithm selects the individuals of the Pareto Front and choose between them using a crowding function, i.e, it penalises closer solutions. If there are not enough solutions in this first selection, the process is repeated with elements that do not belong to the First Pareto Front.

Selection
The selection algorithm must choose between the individuals of the current population and their descendants. According to its criterion, a number of individuals equal to the population size must be selected. The selection algorithm used in this work is called NSGA-II or Non dominated sorting genetic algorithm (Deb, Pratap, Agarwal, & Meyarivan, 2002). This algorithm selects the individuals of the Pareto Front and choose between them using a crowding function, i.e, it penalises closer solutions. If there are not enough solutions in this first selection, the process is repeated with elements that do not belong to the First Pareto Front.

for dataset I have composed of 10 grey and white images
